### DISCUSSION TITLE
TLSV1.3 PskBinderEntry length & cipher suite selection

### DISCUSSION THREAD

**Discussion by jwalch** [openssl/openssl#22665] [Q&A] [answered]
I have read https://github.com/openssl/openssl/issues/5378 & https://github.com/openssl/openssl/issues/20511 and I am still looking for some guidance around what to expect in terms of PSKs and how to correctly implement the callback(s).

I have observed that on the server-side [tls_parse_ctos_psk](https://opengrok-prd.eng.netapp.com/source/s?defs=tls_parse_ctos_psk&project=dev) near the end seems to expect the length of the PskBinderEntry to align with that of the hash function for the suite set in the callback, which must align with that of the negotiated suite further up in the same function.

This seems somewhat at odds with the s_client callback behavior, which seems to default to SHA-256 (and thus a 32-byte PskBinderEntry) even when you force something like TLS_AES_256_GCM_SHA384 via the -ciphersuites parameter. Per the discussion in https://github.com/openssl/openssl/issues/5378 this seems to align with RFC language stating that external PSKs should use SHA-256 unless stated otherwise.

So the issue I'm experiencing is that if I use a PSK with s_client using -ciphersuites TLS_AES_256_GCM_SHA384, there's seemingly nothing I can do in the server-side callback function that will make tls_parse_ctos_psk() happy. If I try to SSL_session_set_cipher(SHA256) because that's what will align with the client's PskBinderEntry, then the session basically gets rejected. If I try to SSL_session_set_cipher(SHA384), then it just blows up later when the 32-byte PskBinderEntry from the client is "too short" for tls_parse_ctos_psk() expectations.

Just to add more color here, there is a nascent protocol out there built on top of TLSv1.3 which tries to convey the desired cipher suite via the identity string. Setting aside the obvious overlap with TLS ClientHello in the first place, the reason I mention this is that it makes me doubtful the RFC "escape clause" ("default to SHA-256 if no such algorithm is defined") is really applicable. The protocol has defined the length of the PSK itself to align with the cipher suite's hash function and also uses the cipher suite's hash function in other places throughout the protocol stack, so it seems very questionable to me that defaulting to SHA-256 as relates to the PskBinderEntry is really what the protocol authors intended. Nonetheless, we are seeing some client implementations exhibiting this behavior.

**Comment by mattcaswell** [comment, marked as answer]
> This seems somewhat at odds with the s_client callback behavior, which seems to default to SHA-256 (and thus a 32-byte PskBinderEntry) even when you force something like TLS_AES_256_GCM_SHA384 via the -ciphersuites parameter. Per the discussion in https://github.com/openssl/openssl/issues/5378 this seems to align with RFC language stating that external PSKs should use SHA-256 unless stated otherwise.

The s_client behaviour is fairy "dumb". There are two ways to specify a TLSv1.3 psk via s_client. You can either use the "-psk" option which supplies the raw PSK data - but gives you no way of associating an MD with that (which means it takes the default route and uses SHA-256). The other way is to use the "-psk_session" option. The input here should be an OpenSSL SSL_SESSION file which gives you the ability to specify both the key and the MD it is associated with (but it is clearly more difficult to create this file than it is to simply pass the psk itself via -psk). Using this second option means you could use an MD other than SHA-256.

Probably we should introduce a new option to s_client (say "-pskmd") which enables you to specify the associated MD more directly.

If the PSK is using SHA-256, but the ciphersuite that gets negotiated uses SHA-384, then you're out of luck. The PSK will not be used.

**Reply by jwalch** [reply]
Thanks for the help, Matt. I actually wasn't immediately concerned with changing s_client's behavior since it happens to mirror that of a different client, but it will assuredly be helpful to me in the future.

Just to be perfectly explicit: you're saying that a client sending just a 32-byte PskBinderEntry AND only permitting TLS_AES_256_GCM_SHA384 is a conflict? Putting it differently, if we assume the externally-established PSK meets the "default to SHA256" criteria and that we must use it, that precludes TLS_AES_256_GCM_SHA384 (and, I suppose, TLS_AEGIS_256_SHA384)?

Reading closer to the end of 4.2.11 in the RFC, I see "Clients MUST verify [...] that the server selected a cipher suite indicating a Hash associated with the PSK [...]" which would seem to support that. 

This makes me even more convinced that the client behavior (setting aside s_client, which you've explained already) of sending a 32-byte PskBinderEntry for a PSK tied to an identity that implies TLS_AES_256_GCM_SHA384 is indeed incorrect. I will also note that the client is sending many other suites in the ClientHello besides the one implied by the single PSK identity, which is also wrong although at least merely in an unnecessarily permissive way. (It even includes several TLSv1.2 suites, even though the supported_versions extension and the higher-level protocol definition itself restrict to only TLSv1.3.) Point being that I'm not too surprised to learn that this PSK-binding aspect of the client's behavior is not RFC-compliant.

**Reply by mattcaswell** [reply]
> Just to be perfectly explicit: you're saying that a client sending just a 32-byte PskBinderEntry AND only permitting TLS_AES_256_GCM_SHA384 is a conflict?

Yes. It's not a protocol violation to do that, but the PSK will never be selected. Instead it will try to fallback to a normal handshake (which will obviously fail if PSK is the only available authentication mechanism).

> Putting it differently, if we assume the externally-established PSK meets the "default to SHA256" criteria and that we must use it, that precludes TLS_AES_256_GCM_SHA384 (and, I suppose, TLS_AEGIS_256_SHA384)?

The "default to SHA256" only applies if no other hash has been specified. It is perfectly possible to have an externally-established PSK explicitly associated with SHA384. But if you don't make that association then it will be SHA256 and any SHA384 based ciphersuite cannot be used with that PSK.

> I will also note that the client is sending many other suites in the ClientHello besides the one implied by the single PSK identity, which is also wrong although at least merely in an unnecessarily permissive way.

It is perfectly acceptable and allowable for a client to say "here's a PSK, but if you can't use that then fallback to a normal handshake and you can use any of this list of ciphersuites". So, it is not necessarily wrong to advertise ciphersuites that are incompatible with the PSK. But if both sides know up front that only a PSK will be acceptable then sending other incompatible ciphersuites is redundant.

**Reply by jwalch** [reply]
> But if both sides know up front that only a PSK will be acceptable [...]

Right, that's the situation. I get that if there were authentication options available then the other suites could possibly come into play.

My concern on the incompatible suites being presented by the client is that that since on the server-side suite selection seems to precede the PSK callback it could cause problems for the TLS_AES_128_GCM_SHA256 case (which otherwise should work)... If we negotiate TLS_AES_256_GCM_SHA384 first then we'll just realize we can't use the PSK down the road and boom "no suitable signature algorithm"... The client even giving us the possibility is setting us up for failure... I guess I can (ab)use SSL_CTX_set_client_hello_cb() to "peek" at the PSK identity and filter out the incompatible suites, but it's an ugly hack that should not be necessary.

> The "default to SHA256" only applies if no other hash has been specified. It is perfectly possible to have an externally-established PSK explicitly associated with SHA384.

Yes, I looked closer at the NVMe-TCP specification and it does appear to actually define the association... my concern was that it was kind of a protocol definition bug and the client's behavior wasn't a strict violation of it even though as per this discussion it shouldn't work anyway...
